Issue #23: Added authentication and authorization documentations (#25)
dotkernel/api-documentation: docs/book/v4/core-features/authentication.md

# Authentication

Authentication is the process by which an identity is presented to the application. It ensures that the entity
making the request has the proper credentials to access the API.

> DotKernel's API identities are delivered to the application from the client through the `Authorization` request

Reviewer: ```suggestion
DotKernel API identities are delivered to the application from the client through the `Authorization` request
```

header.  If it is present, the application tries to find and assign the identity to the application. If it is not presented,

> DotKernel's API assigns a default `guest` identity, represented by an instance of the class

Reviewer: ```suggestion
DotKernel API assigns a default `guest` identity, represented by an instance of the class
```

The client sends a POST request to the `/security/generate-token` with the following parameters:

- `grant_type` = password.
- `client_id` with the client name (from `oauth_clients` table).
- `client_secret` with the client secret (password from `oauth_clients` table for the client).
- `scope` with the scope from `oauth_scopes` table.
- `username` with the user’s username.
- `password` with the user’s password.

```shell
POST /security/generate-token HTTP/1.1
Accept: application/json
Content-Type: application/json
{
"grant_type": "password",
"client_id": "frontend",
"client_secret": "frontend",
"scope": "api",
"username": "test@dotkernel.com",
"password": "dotkernel"
}
```

The server responds with a JSON as follows:

```php
{
"token_type": "Bearer",
"expires_in": 86400,
"access_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9...",
"refresh_token": "def5020087199939a49d0f2f818..."
}
```

Next time when you make a request to the server to an authenticated endpoint, the client should use
the `Authorization` header request.

```shell
GET /users/1 HTTP/1.1
Accept: application/json
Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9...
```

### Refreshing tokens

> DotKernel's API provides the ability to refresh the access token, generating a new one.

Reviewer: ```suggestion
DotKernel API provides the ability to refresh the access token, by generating a new one using the expired access token's `refresh_token`.
```
